### PR TITLE
Add documentation about hosting PowerShell Core

### LINKED ISSUE #2291: Provide a way to host PowerShell in a CoreCLR application
This is follow-up from #2284
## Steps to reproduce

Create a simple coreclr app that hosts powershell thru `Microsoft.PowerShell.SDK`

**Note**: the current function of `Microsoft.PowerShell.SDK` is to provide an ability to compile binary modules that then could be imported and used in PowerShell on any platform. This issue asks a question: should we allow the same nuget package to enable **hosting** scenario?

Program.cs

``` csharp
using System;
using System.Management.Automation;

class Program
{
    static void Main(string[] args)
    {
        PowerShellAssemblyLoadContextInitializer.SetPowerShellAssemblyLoadContext(AppContext.BaseDirectory);

        using (var ps = PowerShell.Create())
        {
        ps.AddScript("Get-Process | Out-String");
        var result = ps.Invoke()[0];
            Console.WriteLine(result.ToString());
        }
    }
}
```

project.json

``` json
{
  "version": "1.0.0-*",
  "buildOptions": {
    "debugType": "portable",
    "emitEntryPoint": true
  },
  "dependencies": {},
  "frameworks": {
    "netcoreapp1.0": {
      "dependencies": {
        "Microsoft.NETCore.App": {
          "type": "platform",
          "version": "1.0.1"
        },
        "Microsoft.PowerShell.SDK" : "1.0.0-alpha10"
      },
      "imports": "dnxcore50"
    }
  }
}
```

nuget.config

``` xml
<?xml version="1.0" encoding="utf-8"?>
<configuration>
  <config>
    <!-- install packages to a custom path -->
    <add key="repositoryPath" value="Packages"/>
  </config>
  <packageSources>
    <clear />
    <add key="nuget.org" value="https://api.nuget.org/v3/index.json" />
    <add key="powershell-core" value="https://powershell.myget.org/F/powershell-core/api/v3/index.json" />
  </packageSources>
</configuration>
```

dotnet restore works fine, dotnet run doesn't
## Expected behavior

You should be able to run and execute
## Actual behavior

On Unix (note that 6.0.0-alpha10 package doesn't have libs compile under `Unix` configuration, so that's expected)

```
Unhandled Exception: System.TypeInitializationException: The type initializer for 'System.Management.Automation.Runspaces.RunspaceFactory' threw an exception. ---> System.DllNotFoundException: Unable to load DLL 'api-ms-win-eventing-provider-l1-1-0.dll': The specified module could not be found.
 (Exception from HRESULT: 0x8007007E)
   at System.Management.Automation.Tracing.EtwActivity.UnsafeNativeMethods.EventActivityIdControl(ActivityControlCode controlCode, Guid& activityId)
   at System.Management.Automation.Tracing.EtwActivity.GetActivityId()
   at System.Management.Automation.Runspaces.RunspaceFactory..cctor()
   --- End of inner exception stack trace ---
   at System.Management.Automation.Runspaces.RunspaceFactory.CreateRunspace()
   at System.Management.Automation.PowerShell.Worker.CreateRunspaceIfNeededAndDoWork(Runspace rsToUse, Boolean isSync)
   at System.Management.Automation.PowerShell.CoreInvokeHelper[TInput,TOutput](PSDataCollection`1 input, PSDataCollection`1 output, PSInvocationSettings settings)
   at System.Management.Automation.PowerShell.CoreInvoke[TInput,TOutput](PSDataCollection`1 input, PSDataCollection`1 output, PSInvocationSettings settings)
   at System.Management.Automation.PowerShell.CoreInvoke[TOutput](IEnumerable input, PSDataCollection`1 output, PSInvocationSettings settings)
   at System.Management.Automation.PowerShell.Invoke(IEnumerable input, PSInvocationSettings settings)
   at System.Management.Automation.PowerShell.Invoke()
   at Program.Main(String[] args)

```

On Windows

```
Unhandled Exception: System.Management.Automation.Runspaces.PSSnapInException: Cannot load Windows PowerShell snap-in Microsoft.PowerShell.Diagnostics because of the following error: Could not load file or assembly 'Microsoft.PowerShell.Commands'. The system cannot find the file specified.
   at System.Management.Automation.Runspaces.RunspaceConfigForSingleShell.LoadMshSnapinAssembly(PSSnapInInfo mshsnapinInfo)
   at System.Management.Automation.Runspaces.RunspaceConfigForSingleShell.LoadPSSnapIn(PSSnapInInfo mshsnapinInfo)
   at System.Management.Automation.Runspaces.RunspaceConfigForSingleShell.LoadPSSnapIns(Collection`1 mshsnapinInfos, PSConsoleLoadException& warning)
   at System.Management.Automation.Runspaces.RunspaceConfigForSingleShell.LoadConsole(PSConsoleLoadException& warning)
   at System.Management.Automation.Runspaces.RunspaceConfigForSingleShell.CreateDefaultConfiguration()
   at System.Management.Automation.Runspaces.RunspaceConfiguration.Create()
   at System.Management.Automation.Runspaces.RunspaceFactory.CreateRunspace(PSHost host)
   at System.Management.Automation.Runspaces.RunspaceFactory.CreateRunspace()
   at System.Management.Automation.PowerShell.Worker.CreateRunspaceIfNeededAndDoWork(Runspace rsToUse, Boolean isSync)
   at System.Management.Automation.PowerShell.CoreInvokeHelper[TInput,TOutput](PSDataCollection`1 input, PSDataCollection`1 output, PSInvocationSettings settings)
   at System.Management.Automation.PowerShell.CoreInvoke[TInput,TOutput](PSDataCollection`1 input, PSDataCollection`1 output, PSInvocationSettings settings)
   at System.Management.Automation.PowerShell.CoreInvoke[TOutput](IEnumerable input, PSDataCollection`1 output, PSInvocationSettings settings)
   at System.Management.Automation.PowerShell.Invoke(IEnumerable input, PSInvocationSettings settings)
   at System.Management.Automation.PowerShell.Invoke()
   at Program.Main(String[] args)
```
## Environment data

``` powershell
> $PSVersionTable
Name                           Value                                                                                                        
----                           -----                                                                                                        
PSVersion                      6.0.0-alpha                                                                                                  
PSEdition                      Core                                                                                                         
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}                                                                                      
BuildVersion                   3.0.0.0                                                                                                      
GitCommitId                    v6.0.0-alpha.10                                                                                              
CLRVersion                                                                                                                                  
WSManStackVersion              3.0                                                                                                          
PSRemotingProtocolVersion      2.3                                                                                                          
SerializationVersion           1.1.0.1                                                                                                      

```

### PR DESCRIPTION
**Pull request by daxian-dbw**
Fix #2291
Add documentation about how to host PowerShell Core in .NET Core application.

### CODE REVIEW

**Comment by daviwil** [comment]
This is really cool!  Quick question, though: does this tie the application to a specific version of PowerShell Core?  Specifically, if I build against Microsoft.PowerShell.SDK of a specific version, is there any way I could get my application to load assemblies from a newer version of PowerShell Core if I have its filesystem path?

I'm asking because I'm re-evaluating whether I could make PowerShell Editor Services a standalone process again, not a module running inside of powershell.exe.

**Comment by daxian-dbw** [comment]
@daviwil When you build your application this way, a specific version (depending on the package version you use) of PowerShell Core assemblies will be published to your publish folder. But your application is not necessarily tied to them. You can just replace them with a newer version of PS Core assemblies, and your application should continue to work as long as the new PS Core assemblies are targeting the same set of .NET Core packages.

There is one problem, though -- we haven't published any powershell core packages for Linux/OSX yet.
PowerShell Core builds separately for Windows and Linux, so the assemblies are different between Windows and Unix platforms. All PowerShell Core packages we published so far contain windows assemblies only ☹️.

There is a workaround. Build your application using the existing powershell core NuGet packages, but replace the powershell assemblies with those from a Linux/OSX package. That hopefully would work.

File: docs/host-powershell/sample-dotnet1.1/Logic/Logic.csproj [hunk 1]
```diff
@@ -0,0 +1,16 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
+
+  <PropertyGroup>
+    <TargetFramework>netstandard1.6</TargetFramework>
```
Comment (iSazonov) [inline docs/host-powershell/sample-dotnet1.1/Logic/Logic.csproj]: Please clarify why "netstandard1.6", not "netcoreapp1.1" as in docs/host-powershell/sample-dotnet1.1/MyApp/MyApp.csproj ?

File: docs/host-powershell/README.md [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+However, when it comes to hosting PowerShell Core, there will be a layer of redirection for the powershell load context to take effect.
+In a .NET Core application, the entry point assembly that contains the `'Main'` method is loaded in the default load context,
+and thus all assemblies referenced by the entry point assembly, implicitly or explicitly, will also be loaded into the default load context.
+
+In order to have the powershell load context to control assembly loading for the execution of an application,
+the business logic code needs to be extracted out of the entry point assembly and put into a different assembly, say `'Logic.dll'`.
+The entry point `'Main'` method shall do one thing only -- let the powershell load context load `'Logic.dll'` and start the execution of the business logic.
+Once the execution starts this way, all further assembly loading requests will be handled by the powershell load context.
+
+So the above example needs to be altered as follows in a .NET Core application:
```
Comment (mirichmo) [inline docs/host-powershell/README.md]: Suggested rewording:
The following example demonstrates hosting PowerShell Core in a .Net Core Application:

File: docs/host-powershell/README.md [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+
+- For `InitializeAndCallEntryMethod`,
+it's designed to be used with `dotnet.exe` where the TPA list includes powershell assemblies.
+When using this API, `PowerShellAssemblyLoadContextInitializer` will set up a new load context to handle all assemblies.
+PowerShell Core itself also uses this API for [bootstrapping][].
+
+This documentation only covers the `InitializeAndCallEntryMethod` API,
+as it's what you need when building a .NET Core application with .NET CLI.
+
+### Example
```
Comment (mirichmo) [inline docs/host-powershell/README.md]: There should be three sub-headings here:
1. Hosting Windows PowerShell
2. Hosting PowerShell in .Net Core
3. Example .Net Core program

**Review by mirichmo — DISMISSED** [review]

**Comment by daviwil** [comment]
@daxian-dbw thanks for the explanation!  Seems like it would be more complex go have my own process hosting PowerShell so I think I'll stick with my current model of being hosted inside of powershell.exe :)

File: docs/host-powershell/sample-dotnet1.1/Logic/Logic.csproj (hunk 1 above)
Comment (daxian-dbw) [inline docs/host-powershell/sample-dotnet1.1/Logic/Logic.csproj]: Becuase `Logic.dll` is an assembly Library while `MyApp.exe` is an executable. If you inspect the dependencies of `Microsoft.NetCore.App`, you will see that it depends on `NetStandard.Library` and also include the runtime and host related packages.

File: docs/host-powershell/README.md [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+
+Due to the lack of necessary APIs for manipulating assemblies in .NET Core 1.1 and prior,
+PowerShell Core needs to control assembly loading via our customized `AssemblyLoadContext` in order to do tasks like type resolution.
+So applications that want to host PowerShell Core (using PowerShell APIs) need to be bootstrapped from `PowerShellAssemblyLoadContextInitializer`.
+
+`PowerShellAssemblyLoadContextInitializer` exposes 2 APIs for this purpose:
+`SetPowerShellAssemblyLoadContext` and `InitializeAndCallEntryMethod`.
+They are for different scenarios:
+
+- For `SetPowerShellAssemblyLoadContext`,
```
Comment (daxian-dbw) [inline docs/host-powershell/README.md]: Fixed.

File: docs/host-powershell/README.md [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+it's designed to be used by a native host whose TPA (Trusted Platform Assemblies) list doesn't include powershell assemblies,
```
Comment (daxian-dbw) [inline docs/host-powershell/README.md]: Fixed.

File: docs/host-powershell/README.md [hunk 6]
```diff
@@ -0,0 +1,149 @@
...
+When using this API, instead of setting up a new load context,
+`PowerShellAssemblyLoadContextInitializer` will register a handler to the [Resolving][] event of the default load context.
+Then PowerShell Core will depend on the default load context to handle TPA and the `Resolving` event to handle other assemblies.
+
+- For `InitializeAndCallEntryMethod`,
+it's designed to be used with `dotnet.exe` where the TPA list includes powershell assemblies.
+When using this API, `PowerShellAssemblyLoadContextInitializer` will set up a new load context to handle all assemblies.
+PowerShell Core itself also uses this API for [bootstrapping][].
+
+This documentation only covers the `InitializeAndCallEntryMethod` API,
```
Comment (daxian-dbw) [inline docs/host-powershell/README.md]: We own those 2 APIs. They are exposed from `PowerShellAssemblyLoadContextInitializer`.

File: docs/host-powershell/README.md (hunk 3 above)
Comment (daxian-dbw) [inline docs/host-powershell/README.md]: Fixed.

File: docs/host-powershell/README.md [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+as it's what you need when building a .NET Core application with .NET CLI.
+
+### Example
+
+The following code is how you host Windows PowerShell in an application.
```
Comment (daxian-dbw) [inline docs/host-powershell/README.md]: I don't intend to introduce how to host Windows PowerShell in full .NET application.  
The code about hosting Windows PowerShell is used as a comparison to show what you need to do in .NET Core for the same purpose.

File: docs/host-powershell/README.md (hunk 2 above)
Comment (daxian-dbw) [inline docs/host-powershell/README.md]: Same here. The code about hosting windows powershell is used as a comparison to show how this familiar code should be altered when it comes to .NET Core 1.1.

File: docs/host-powershell/sample-dotnet1.1/Logic/Logic.csproj (hunk 1 above)
Comment (iSazonov) [inline docs/host-powershell/sample-dotnet1.1/Logic/Logic.csproj]: Thanks!

File: docs/host-powershell/README.md (hunk 7 above)
Comment (daxian-dbw) [inline docs/host-powershell/README.md]: But I reworded it as you suggested.

**Comment by daxian-dbw** [comment]
@mirichmo I addressed your comments. Could you please take another look?

File: docs/host-powershell/README.md [hunk 8]
```diff
@@ -0,0 +1,152 @@
...
+Due to the lack of necessary APIs for manipulating assemblies in .NET Core 1.1 and prior,
+PowerShell Core needs to control assembly loading via our customized `AssemblyLoadContext` ([CorePsAssemblyLoadContext.cs][]) in order to do tasks like type resolution.
+So applications that want to host PowerShell Core (using PowerShell APIs) need to be bootstrapped from `PowerShellAssemblyLoadContextInitializer`.
+
+`PowerShellAssemblyLoadContextInitializer` exposes 2 APIs for this purpose:
+`SetPowerShellAssemblyLoadContext` and `InitializeAndCallEntryMethod`.
+They are for different scenarios:
+
+- `SetPowerShellAssemblyLoadContext` - It's designed to be used by a native host
+whose Trusted Platform Assemblies (TPA) do not include powershell assemblies,
```
Comment (joeyaiello) [inline docs/host-powershell/README.md]: Properly case "PowerShell"

File: docs/host-powershell/README.md [hunk 9]
```diff
@@ -0,0 +1,152 @@
...
+such as the in-box `powershell.exe` and other native CoreCLR host in Nano Server.
+When using this API, instead of setting up a new load context,
+`PowerShellAssemblyLoadContextInitializer` will register a handler to the [Resolving][] event of the default load context.
+Then PowerShell Core will depend on the default load context to handle TPA and the `Resolving` event to handle other assemblies.
+
+- `InitializeAndCallEntryMethod` - It's designed to be used with `dotnet.exe`
+where the TPA list includes powershell assemblies.
```
Comment (joeyaiello) [inline docs/host-powershell/README.md]: Properly case "PowerShell"

File: docs/host-powershell/README.md [hunk 10]
```diff
@@ -0,0 +1,152 @@
...
+When using this API, `PowerShellAssemblyLoadContextInitializer` will set up a new load context to handle all assemblies.
+PowerShell Core itself also uses this API for [bootstrapping][].
+
+This documentation only covers the `InitializeAndCallEntryMethod` API,
+as it's what you need when building a .NET Core application with .NET CLI.
+
+### Comparison - Hosting Windows PowerShell vs. Hosting PowerShell Core
+
+The following code demonstrates how to host Windows PowerShell in an application.  
+As shown below, you can insert your business logic code directly in the `'Main'` method.
```
Comment (joeyaiello) [inline docs/host-powershell/README.md]: No reason to use both backticks and single quotes. Backticks will do here.

File: docs/host-powershell/README.md [hunk 11]
```diff
@@ -0,0 +1,152 @@
...
+        using (PowerShell ps = PowerShell.Create())
+        {
+            var results = ps.AddScript("Get-Command Write-Output").Invoke();
+            Console.WriteLine(results[0].ToString());
+        }
+    }
+}
+```
+
+However, when it comes to hosting PowerShell Core, there will be a layer of redirection for the powershell load context to take effect.
```
Comment (joeyaiello) [inline docs/host-powershell/README.md]: "PowerShell" casing

File: docs/host-powershell/README.md [hunk 12]
```diff
@@ -0,0 +1,152 @@
...
+In a .NET Core application, the entry point assembly that contains the `'Main'` method is loaded in the default load context,
```
Comment (joeyaiello) [inline docs/host-powershell/README.md]: No single quotes needed

File: docs/host-powershell/README.md [hunk 13]
```diff
@@ -0,0 +1,152 @@
...
+and thus all assemblies referenced by the entry point assembly, implicitly or explicitly, will also be loaded into the default load context.
+
+In order to have the powershell load context to control assembly loading for the execution of an application,
```
Comment (joeyaiello) [inline docs/host-powershell/README.md]: PowerShell casing

File: docs/host-powershell/README.md [hunk 14]
```diff
@@ -0,0 +1,152 @@
...
+the business logic code needs to be extracted out of the entry point assembly and put into a different assembly, say `'Logic.dll'`.
```
Comment (joeyaiello) [inline docs/host-powershell/README.md]: No single quotes

File: docs/host-powershell/README.md [hunk 15]
```diff
@@ -0,0 +1,152 @@
...
+The entry point `'Main'` method shall do one thing only -- let the powershell load context load `'Logic.dll'` and start the execution of the business logic.
```
Comment (joeyaiello) [inline docs/host-powershell/README.md]: No single quotes

File: docs/host-powershell/README.md [hunk 16]
```diff
@@ -0,0 +1,152 @@
...
+
+### .NET Core Sample Application
+
+You can find the sample application project `"MyApp"` under [sample-dotnet1.1](./sample-dotnet1.1). 
+To build the sample project, run the following commands:
+
+```powershell
+dotnet restore .\MyApp\MyApp.csproj
+dotnet publish .\MyApp -c release -r win10-x64
+```
```
Comment (joeyaiello) [inline docs/host-powershell/README.md]: Has this document assumed that the user has installed dotnet-cli? Might it be worth putting a pointer to environment instructions (preferably not owned by us) at the top of the document?

Comment (joeyaiello) [inline docs/host-powershell/README.md]: Also, is this the same on Linux with a different RID? Might that be worth calling out?

**Review by joeyaiello — CHANGES_REQUESTED** [review]
Just a few small changes, otherwise looks really useful to me.

File: docs/host-powershell/README.md (hunk 8 above)
Comment (daxian-dbw) [inline docs/host-powershell/README.md]: Fixed.

File: docs/host-powershell/README.md (hunk 9 above)
Comment (daxian-dbw) [inline docs/host-powershell/README.md]: Fixed.

File: docs/host-powershell/README.md (hunk 10 above)
Comment (daxian-dbw) [inline docs/host-powershell/README.md]: Fixed.

File: docs/host-powershell/README.md (hunk 11 above)
Comment (daxian-dbw) [inline docs/host-powershell/README.md]: Fixed.

File: docs/host-powershell/README.md (hunk 12 above)
Comment (daxian-dbw) [inline docs/host-powershell/README.md]: Fixed.

File: docs/host-powershell/README.md (hunk 13 above)
Comment (daxian-dbw) [inline docs/host-powershell/README.md]: Fixed.

File: docs/host-powershell/README.md (hunk 14 above)
Comment (daxian-dbw) [inline docs/host-powershell/README.md]: Fixed.

File: docs/host-powershell/README.md (hunk 15 above)
Comment (daxian-dbw) [inline docs/host-powershell/README.md]: Fixed.

File: docs/host-powershell/README.md (hunk 16 above)
Comment (daxian-dbw) [inline docs/host-powershell/README.md]: Unfortunately, as [discussed above](https://github.com/PowerShell/PowerShell/pull/3409#issuecomment-288934991), all powershell NuGet packages that have been published so far only contain powershell assemblies built specifically for Windows. We need to publish NuGet packages that contain both Windows and Unix assemblies, but `dotnet pack` doesn't seem to support that.

#3417 is opened to track this.

**Comment by daxian-dbw** [comment]
@joeyaiello @mirichmo All comments have been addressed. Can you please take another look?

**Review by joeyaiello — APPROVED** [review]
LGTM, thanks!

**Review by mirichmo — APPROVED** [review]

**Pull request merged**